jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/ProjectCommand.java

```diff
@@ -79,4 +96,79 @@ private static String getWorkspaceInvisibleProjectName(IPath workspacePath) {
         String fileName = workspacePath.toFile().getName();
         return fileName + "_" + Integer.toHexString(workspacePath.toPortableString().hashCode());
     }
+
+    public static List<PackageNode> getMainMethod(IProgressMonitor monitor) {
+        final List<PackageNode> res = new ArrayList<>();
+        IJavaSearchScope scope = SearchEngine.createWorkspaceScope();
+        SearchPattern pattern = SearchPattern.createPattern("main(String[]) void", IJavaSearchConstants.METHOD,
+                IJavaSearchConstants.DECLARATIONS, SearchPattern.R_EXACT_MATCH| SearchPattern.R_CASE_SENSITIVE);
+        SearchRequestor requestor = new SearchRequestor() {
+            @Override
+            public void acceptSearchMatch(SearchMatch match) {
+                Object element = match.getElement();
+                if (element instanceof IMethod) {
+                    IMethod method = (IMethod) element;
+                    try {
+                        if (method.isMainMethod()) {
+                            IResource resource = method.getResource();
+                            if (resource != null) {
+                                IProject project = resource.getProject();
+                                if (project != null) {
+                                    String mainClass = method.getDeclaringType().getFullyQualifiedName();
+                                    IJavaProject javaProject = getJavaProject(project);
+                                    if (javaProject != null) {
+                                        String moduleName = getModuleName(javaProject);
+                                        if (moduleName != null) {
+                                            mainClass = moduleName + "/" + mainClass;
+                                        }
+                                    }
+                                    String projectName = ProjectsManager.DEFAULT_PROJECT_NAME.equals(project.getName()) ? null : project.getName();
+                                    String filePath = null;
+                                    if (match.getResource() instanceof IFile) {
+                                        try {
+                                            filePath = match.getResource().getLocation().toOSString();
+                                        } catch (Exception ex) {
+                                                // ignore
+                                        }
+                                    }
+                                    res.add(new PackageNode(mainClass, filePath, NodeKind.FILE));
```

Comment:
It is only used as a data struct to storage data from server. Is it better to use a standalone class?

Reply:
> Is there any special reason why we need to use `PackageNode` here? From my understanding, it's for the tree view.

It is only used as a data struct to storage data from server. Is it better to use a standalone class?

Reply:
Using `PackageNode` will save some code, but it will make the code harder to read and maintain in the future. I suggest to use a new data structure here